lib/buildImage.nix

```diff
@@ -0,0 +1,133 @@
+  #   regex = ".*";
+  #   mode = "0664";
+  # }
+  # The mode is applied on a specific path. In this path subtree,
+  # the mode is then applied on all files matching the regex.
+  perms ? [ ]
+, # The maximun number of layer to create. This is based on the
+  # store path "popularity" as described in
+  # https://grahamc.com/blog/nix-and-layered-docker-images
+  # Note this is applied on the image layers and not on layers added
+  # with the buildImage.layers attribute
+  maxLayers ? 1
+, # If set to true, the Nix database is initialized with all store
+  # paths added into the image. Note this is only useful to run nix
+  # commands from the image, for instance to build an image used by
+  # a CI to run Nix builds.
+  initializeNixDatabase ? false
+, # If initializeNixDatabase is set to true, the uid/gid of /nix can be
+  # controlled using nixUid/nixGid.
+  nixUid ? 0
+, nixGid ? 0
+, # Deprecated: will be removed
+  contents ? null
+, meta ? { }
+,
+}:
+let
+  configFile = writeText "config.json" (builtins.toJSON config);
+  copyToRootList =
+    let derivations = if !isNull contents then contents else copyToRoot;
+    in if isNull derivations
+    then [ ]
+    else if !builtins.isList derivations
+    then [ derivations ]
+    else derivations;
+
+  # Expand the given list of layers to include all their transitive layer dependencies.
+  layersWithNested = layers:
+    let layerWithNested = layer: [ layer ] ++ (builtins.concatMap layerWithNested (layer.layers or [ ]));
+    in builtins.concatMap layerWithNested layers;
+  explodedLayers = layersWithNested layers;
+  ignore = [ configFile ] ++ explodedLayers;
+
+  closureGraphForAllLayers = closureGraph ([ configFile ] ++ copyToRootList ++ layers) ignore;
+  nixDatabase = makeNixDatabase closureGraphForAllLayers;
+  # This layer contains all config dependencies. We ignore the
+  # configFile because it is already part of the image, as a
+  # specific blob.
+
+  perms' = perms ++ lib.optionals initializeNixDatabase
+    [
+      {
+        path = nixDatabase;
+        regex = ".*";
+        mode = "0755";
+        uid = nixUid;
+        gid = nixGid;
+      }
+    ];
+
+  customizationLayer = buildLayer {
+    inherit maxLayers;
+    perms = perms';
+    copyToRoot =
+      if initializeNixDatabase
+      then copyToRootList ++ [ nixDatabase ]
+      else copyToRootList;
+    deps = [ configFile ];
+    ignore = configFile;
+    layers = layers;
+  };
+  fromImageFlag = lib.optionalString (fromImage != "") "--from-image ${fromImage}";
+  layerPaths = lib.concatMapStringsSep " " (l: l + "/layers.json") (layers ++ [ customizationLayer ]);
+  image =
+    let
+      imageName = lib.toLower name;
+      imageTag =
+        if tag != null
+        then tag
+        else
+          lib.head (lib.strings.splitString "-" (baseNameOf image.outPath));
+    in
+    runCommand "image-${baseNameOf name}.json"
+      {
+        inherit imageName meta;
+        nativeBuildInputs = [ nix2container-bin ];
```

Comment:
This is the core change across most derivations. We reference `nix2container-bin` through `nativeBuildInputs` instead of `${nix2container-bin}/bin/nix2container`. Paired with `makeScopeWithSplicing`/`callPackage` this automatically resolves to the correct variant of the package.